### PR TITLE
Implement basic features of Hearthstone: Battlegrounds

### PR DESCRIPTION
**Pull request by utilForever**
This revision includes:
- Create a new mode: Hearthstone Battlegrounds (#349)
  - This mode is implemented w/o pointers. Instead, I used modern C++ features such as `std::variant`, `std::visit`, `std::function`. Also, I used `std::array` instead of `std::vector` if possible.
  - Support features
    - Select hero
    - Purchase a minion in the Tavern's field
    - Sell a minion in the player's field
    - Rearrange a list of minions in the player's field
    - Refresh a list of minions in the Tavern's field
    - Upgrade the player's Tavern to the next tier

### CODE REVIEW

File: Sources/Rosetta/Battlegrounds/Cards/Card.cpp [hunk 1]
```diff
@@ -0,0 +1,202 @@
...
+    // Red Whelp (59968)
+    // Rockpool Hunter (41245)
+    // Selfless Hero (38740)
+    // Vulgar Homunculus (43121)
+    // Wrath Weaver (59670)
+    if (dbfID == 40426 || dbfID == 985 || dbfID == 60628 || dbfID == 56112 ||
+        dbfID == 48886 || dbfID == 60055 || dbfID == 475 || dbfID == 976 ||
+        dbfID == 42467 || dbfID == 59968 || dbfID == 41245 || dbfID == 38740 ||
+        dbfID == 43121 || dbfID == 59670)
+    {
```
Comment (kozistr) [inline Sources/Rosetta/Battlegrounds/Cards/Card.cpp:47]: 요 부분은 const array(?) 같은 거로 만들어 두고 iterate 하면서 checking 해 보면 어떨까용? 유지보수 측에서 더 좋을 것 같아서요!

File: Sources/Rosetta/Battlegrounds/Cards/Card.cpp [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+    // Spawn of N'Zoth (38797)
+    // Steward of Time (60621)
+    // Unstable Ghoul (1808)
+    // Waxrider Togwaggle (60559)
+    // Zoobot (39839)
+    if (dbfID == 61029 || dbfID == 778 || dbfID == 59937 || dbfID == 49279 ||
+        dbfID == 39481 || dbfID == 2016 || dbfID == 1063 || dbfID == 59186 ||
+        dbfID == 736 || dbfID == 60122 || dbfID == 40428 || dbfID == 1281 ||
+        dbfID == 38797 || dbfID == 60621 || dbfID == 1808 || dbfID == 60559 ||
+        dbfID == 39839)
```
Comment (kozistr) [inline Sources/Rosetta/Battlegrounds/Cards/Card.cpp:78]: here too

File: Sources/Rosetta/Battlegrounds/Cards/Card.cpp [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+    // Screwjank Clunker (2023)
+    // Shifter Zerus (57742)
+    // Soul Juggler (59660)
+    // The Beast (962)
+    // Twilight Emissary (60626)
+    if (dbfID == 60558 || dbfID == 453 || dbfID == 40391 || dbfID == 2518 ||
+        dbfID == 61930 || dbfID == 56393 || dbfID == 60552 || dbfID == 1003 ||
+        dbfID == 2288 || dbfID == 38734 || dbfID == 52502 || dbfID == 59940 ||
+        dbfID == 60048 || dbfID == 48536 || dbfID == 2023 || dbfID == 57742 ||
+        dbfID == 59660 || dbfID == 962 || dbfID == 60626)
```
Comment (kozistr) [inline Sources/Rosetta/Battlegrounds/Cards/Card.cpp:112]: here too

File: Sources/Rosetta/Battlegrounds/Cards/Card.cpp [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+    // Menagerie Magician (39269)
+    // Savannah Highmane (1261)
+    // Security Rover (48100)
+    // Siegebreaker (54835)
+    // Toxfin (52277)
+    // Virmen Sensei (40641)
+    if (dbfID == 48993 || dbfID == 45392 || dbfID == 43358 || dbfID == 42442 ||
+        dbfID == 763 || dbfID == 61072 || dbfID == 2068 || dbfID == 60498 ||
+        dbfID == 1992 || dbfID == 49169 || dbfID == 39269 || dbfID == 1261 ||
+        dbfID == 48100 || dbfID == 54835 || dbfID == 52277 || dbfID == 40641)
```
Comment (kozistr) [inline Sources/Rosetta/Battlegrounds/Cards/Card.cpp:142]: here too

File: Sources/Rosetta/Battlegrounds/Cards/Card.cpp [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+    // Murozond (60637)
+    // Primalfin Lookout (60028)
+    // Razorgore, the Untamed (60561)
+    // Sneed's Old Shredder (59682)
+    // Strongshell Scavenger (43022)
+    // Voidlord (46056)
+    if (dbfID == 59714 || dbfID == 1915 || dbfID == 2949 || dbfID == 59955 ||
+        dbfID == 49973 || dbfID == 2074 || dbfID == 60247 || dbfID == 59707 ||
+        dbfID == 1986 || dbfID == 60637 || dbfID == 60028 || dbfID == 60561 ||
+        dbfID == 59682 || dbfID == 43022 || dbfID == 46056)
```
Comment (kozistr) [inline Sources/Rosetta/Battlegrounds/Cards/Card.cpp:171]: here too

File: Sources/Rosetta/Battlegrounds/Cards/Card.cpp [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+    // Holy Mackerel (61079)
+    // Imp Mama (61028)
+    // Kalecgos, Arcane Aspect (60630)
+    // Maexxna (1791)
+    // Mama Bear (60036)
+    // Nadina the Red (60629)
+    // Zapp Slywick (60040)
+    if (dbfID == 2081 || dbfID == 59935 || dbfID == 56465 || dbfID == 52041 ||
+        dbfID == 61079 || dbfID == 61028 || dbfID == 60630 || dbfID == 1791 ||
+        dbfID == 60036 || dbfID == 60629 || dbfID == 60040)
```
Comment (kozistr) [inline Sources/Rosetta/Battlegrounds/Cards/Card.cpp:195]: here too

File: Sources/Rosetta/Battlegrounds/Cards/Card.cpp (hunk 1 above)
Comment (utilForever) [reply Sources/Rosetta/Battlegrounds/Cards/Card.cpp]: Fixed.

File: Sources/Rosetta/Battlegrounds/Cards/Card.cpp (hunk 2 above)
Comment (utilForever) [reply Sources/Rosetta/Battlegrounds/Cards/Card.cpp]: Fixed.

File: Sources/Rosetta/Battlegrounds/Cards/Card.cpp (hunk 3 above)
Comment (utilForever) [reply Sources/Rosetta/Battlegrounds/Cards/Card.cpp]: Fixed.

File: Sources/Rosetta/Battlegrounds/Cards/Card.cpp (hunk 4 above)
Comment (utilForever) [reply Sources/Rosetta/Battlegrounds/Cards/Card.cpp]: Fixed.

File: Sources/Rosetta/Battlegrounds/Cards/Card.cpp (hunk 5 above)
Comment (utilForever) [reply Sources/Rosetta/Battlegrounds/Cards/Card.cpp]: Fixed.

File: Sources/Rosetta/Battlegrounds/Cards/Card.cpp (hunk 6 above)
Comment (utilForever) [reply Sources/Rosetta/Battlegrounds/Cards/Card.cpp]: Fixed.

**Pull request merged**